runtime/object.go

```diff
@@ -182,6 +185,89 @@ func NewInt(s string) Int {
 	return Int{i}
 }
 
+type EnumRange struct {
+	First, Last int
+}
+
+func (er *EnumRange) Contains(x int) bool {
+	return x >= er.First && x <= er.Last
+}
+func (r EnumRange) ToString() string {
+	if r.First == r.Last {
+		return fmt.Sprintf("%d", r.First)
+	}
+	return fmt.Sprintf("%d..%d", r.First, r.Last)
+}
+
+type EnumType struct {
+	Name     string
+	Elements map[string][]EnumRange
+}
+
+func (e *EnumType) Type() ObjectType { return ENUMTYPE }
+func (e *EnumType) Inspect() string {
+	var ret []string
+	for name, val := range e.Elements {
+		var retE []string
+		for _, r := range val {
+			retE = append(retE, r.ToString())
+		}
+		ret = append(ret, fmt.Sprintf("%s(%s)", name, strings.Join(retE, ", ")))
+	}
+	return e.Name + "{" + strings.Join(ret, ", ") + "}"
+}
+
+func (e *EnumType) Equal(obj Object) bool {
+	other, ok := obj.(*EnumType)
+	if !ok {
+		return false
+	}
+	return reflect.DeepEqual(e, other)
+}
+
+func NewEnumType(enumTypeName string, Enums ...string) *EnumType {
+	ret := EnumType{}
+	ret.Name = enumTypeName
+	ret.Elements = make(map[string][]EnumRange)
+	for EnumId, EnumName := range Enums {
+		ranges := []EnumRange{{First: EnumId, Last: EnumId}}
+		ret.Elements[EnumName] = ranges
+	}
+	return &ret
+}
+
+type Enum struct {
+	typeRef *EnumType
+	key     string
+	frozen  bool
```

Comment:
I have not understood what `frozen` is for. Would you mind explaining it again?

Reply:
done: 'frozen' removed

while evaluating new enum type, separate runtime object are added for each key

example:
type enumerated E4 {red, blue}

will add 2 runtime object
1. 'red'
2. 'blue'

'frozen' was used to mark those objects